### PR TITLE
[spirv] sort debug instructions to remove invalid forward references

### PR DESCRIPTION
**Pull request by jaebaek**
This commit uses post-order travel for the topological sort.

### CODE REVIEW

**Comment by jaebaek** [comment]
@dnovillo, could you please review this commit?

File: tools/clang/lib/SPIRV/CMakeLists.txt [hunk 1]
```diff
@@ -40,3 +40,4 @@ add_clang_library(clangSPIRV
 
 target_include_directories(clangSPIRV PUBLIC ${SPIRV_HEADER_INCLUDE_DIR})
 target_include_directories(clangSPIRV PRIVATE ${SPIRV_TOOLS_INCLUDE_DIR})
+add_dependencies(clangSPIRV TablegenHLSLOptions)
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/CMakeLists.txt:44]: why is this needed?

File: tools/clang/lib/SPIRV/EmitVisitor.cpp [hunk 2]
```diff
@@ -341,8 +341,9 @@ void EmitVisitor::encodeString(llvm::StringRef value) {
   curInst.insert(curInst.end(), words.begin(), words.end());
 }
 
-bool EmitVisitor::visit(SpirvModule *, Phase) {
-  // No pre-visit operations needed for SpirvModule.
+bool EmitVisitor::visit(SpirvModule *mod, Phase phase) {
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/EmitVisitor.cpp]: I recommend creating a separate Visitor class (e.g. SortDebugInfoVisitor) and put this logic in that visitor instead of adding it to EmitVisitor.

The main reason for this recommendation is the following:
By having this logic in `EmitVisitor` all other passes (e.g. LowerTypeVisitor, CapabilityVisitor, etc.) see the debug instructions in a different (wrong) order. And the order is only fixed right before emitting. This may be okay right now because other passes are unlikely to do anything with debug instructions, but it could change in the future.

I also think that `sortDebugInstructionsInPostOrder` and `whileEachOperandOfDebugInstruction` should be member functions of `SortDebugInfoVisitor`, instead of being member function of the module.

What do you think?

File: tools/clang/lib/SPIRV/SpirvModule.cpp [hunk 3]
```diff
@@ -459,63 +249,188 @@ void SpirvModule::addSource(SpirvSource *src) {
...
+      return;
+
+  switch (di->getKind()) {
+  case SpirvInstruction::IK_DebugCompilationUnit: {
+    SpirvDebugCompilationUnit *inst = dyn_cast<SpirvDebugCompilationUnit>(di);
+    assert(inst != nullptr);
+    if (!visitor(inst->getDebugSource()))
+      return;
+  }
+    return;
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/SpirvModule.cpp]: You should use `break;` instead of having `return;` after every `case` statement. The [Google C++ style guide](https://google.github.io/styleguide/cppguide.html#Loops_and_Switch_Statements) suggests:
```cpp
switch (var) {
  case 0: {
    ...
    break;
  }
  case 1: {
    ...
    break;
  }
  default: {
    assert(false);
  }
}
```

File: tools/clang/lib/SPIRV/SpirvModule.cpp [hunk 4]
```diff
@@ -459,63 +249,188 @@ void SpirvModule::addSource(SpirvSource *src) {
...
+    SpirvDebugTypeTemplateParameter *inst =
+        dyn_cast<SpirvDebugTypeTemplateParameter>(di);
+    assert(inst != nullptr);
+    if (!visitor(inst->getActualType()))
+      return;
+    if (!visitor(inst->getSource()))
+      return;
+  }
+    return;
+  default:
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/SpirvModule.cpp]: If a debug instruction is added in the future, and it's missing from this switch statement, that would be a potential bug that we won't know about. Please add an `assert(false && "unsupported debug instruction");`

File: tools/clang/lib/SPIRV/SpirvModule.cpp [hunk 5]
```diff
@@ -459,63 +249,188 @@ void SpirvModule::addSource(SpirvSource *src) {
...
+  }
 
-void SpirvModule::addModuleProcessed(SpirvModuleProcessed *p) {
-  assert(p);
-  moduleProcesses.push_back(p);
+  // Sort debug instructions in a post order. We puts successors in the first
+  // places of `debugInstructions`. For example, `DebugInfoNone` does not have
+  // any operand, which means it does not have any successors. We have to place
+  // it earlier than the instructions using it.
+  debugInstructions.clear();
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/SpirvModule.cpp]: You are clearing a vector of pointers and then re-populating the vector. I highly suggest getting the size of `debugInstructions` before clearing it, and then after the while loop add an `assert` to make sure the new size is equal to the old size. Otherwise you have lost pointers to memory forever (leak).

In addition to leaks, not having the assert could result in really-hard-to-find bugs. For example if `whileEachOperandOfDebugInstruction` is missing a debug instruction `case` statement or if it is missing an operand of some debug instruction, you could have a leak, which can be caught by the assert.

**Review by ehsannas — CHANGES_REQUESTED** [review]
Nice job. In general I really like this change.
Very happy about:
- Removing all the case-specific member functions in SpirvModule
- Removing custom visiting (invokeVisitorDebugLexicalScope)

I have a few recommendations pointed out below.

File: tools/clang/lib/SPIRV/CMakeLists.txt (hunk 1 above)
Comment (jaebaek) [reply tools/clang/lib/SPIRV/CMakeLists.txt:44]: I do not think it is related to this change but keeping failing in the AppVeyor test motivates me to add this dependency.

As far as I remember, I added `#include "dxc/Support/HLSLOptions.h"` in `tools/clang/lib/SPIRV/EmitVisitor.cpp` for reading the source code to dump "not preprocessed" code. The AppVeyor test fails because of missing `HLSLOptions.inc` which is needed by `#include "dxc/Support/HLSLOptions.h"` in `tools/clang/lib/SPIRV/EmitVisitor.cpp`. `TablegenHLSLOptions` target generates `HLSLOptions.inc`. It will resolve the flaky test failure.

File: tools/clang/lib/SPIRV/EmitVisitor.cpp (hunk 2 above)
Comment (jaebaek) [reply tools/clang/lib/SPIRV/EmitVisitor.cpp]: I agree that creating a separate visitor will be better. It will help us not forget that we sort the instructions before emission.

In terms of `sortDebugInstructionsInPostOrder` and `whileEachOperandOfDebugInstruction`, I was worried about updating the vector for debug instructions that will be entire drops of existing instructions and adding new instructions. However, since I do not think we will have a large size of the vector, it might be ok. I agree with you for it as well. Let me update the code.

File: tools/clang/lib/SPIRV/SpirvModule.cpp (hunk 3 above)
Comment (jaebaek) [reply tools/clang/lib/SPIRV/SpirvModule.cpp]: done

File: tools/clang/lib/SPIRV/SpirvModule.cpp (hunk 4 above)
Comment (jaebaek) [reply tools/clang/lib/SPIRV/SpirvModule.cpp]: done

File: tools/clang/lib/SPIRV/SpirvModule.cpp (hunk 5 above)
Comment (jaebaek) [reply tools/clang/lib/SPIRV/SpirvModule.cpp]: Oh nice suggestion

**Review by jaebaek — COMMENTED** [review]
Thank you for your review! Updated based on comments. PTAL!

**Review by ehsannas — APPROVED** [review]
\o/ Thanks

**Comment by jaebaek** [comment]
Thanks!
